libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BaseDownloadSyncAdapter.kt

```diff
@@ -0,0 +1,249 @@
+
+                for ((key, value) in queryKnownObjects(ids)) {
+                    prepareDelete(value)
+                    delete(key)
+                }
+            }
+        } catch (e: ResourceNotModified) {
+            // Do nothing
+        }
+    }
+
+    protected open fun getKnownCount(): Long = knownIDs.let {
+        if (it == null) {
+            throw RuntimeException("knownIDs can't be null if deleteUnseen() returns true.")
+        }
+        return it.size.toLong()
+    }
+
+    protected open fun queryKnownObjects(ids: Set<K>): MutableMap<K, T> {
+        if (ids.isEmpty()) {
+            return mutableMapOf()
+        }
+
+        val it = BatchedQueryIterator(ids.iterator(), this)
+        val known = mutableMapOf<K, T>()
+        while (it.hasNext()) {
+            try {
+                val obj = it.next()
+                val key = getId(obj)
+                if (known.containsKey(key)) {
+                    delete(key)
+                }
+                known[key] = obj
+            } catch (e: BatchEmptyException) {
+                // Ignore
+            }
+        }
+        return known
+    }
+
+    protected open fun preprocessObject(obj: JSONObject): JSONObject {
+        return obj
+    }
+
+    protected fun processPage(data: JSONArray) {
+        val l = data.length()
+        runInTransaction {
+            val fetchIds: MutableSet<K> = HashSet()
+            for (i in 0 until l) {
+                val jsonobj = data.getJSONObject(i)
+                fetchIds.add(getId(jsonobj))
+            }
+            val known: MutableMap<K, T> = queryKnownObjects(fetchIds)
+            for (i in 0 until l) {
+                val jsonobj = preprocessObject(data.getJSONObject(i))
+                jsonobj.put(
+                    "__libpretixsync_dbversion",
+                    Migrations.CURRENT_VERSION
+                )
+                jsonobj.put("__libpretixsync_syncCycleId", syncCycleId)
+                val jsonid: K = getId(jsonobj)
+                var obj: T?
+                var old: JSONObject? = null
+
+                if (seenIDs.contains(jsonid)) {
+                    continue
+                } else if (known.containsKey(jsonid)) {
+                    obj = known.getValue(jsonid)
+                    old = getJSON(obj)
+                } else {
+                    obj = null
+                }
+                if (obj != null) {
+                    known.remove(jsonid)
+                    knownIDs?.let {
+                        it.remove(jsonid)
+                    }
+                    if (!JSONUtils.similar(jsonobj, old)) {
+                        update(obj, jsonobj)
+                    }
+                } else {
+                    // TODO: Is it okay to insert directly instead of batching?
+                    if (autoPersist()) {
+                        insert(jsonobj)
+                        inserted += 1
+                    }
```

Comment:
Since we do no batching any more, autoPersist() does not make much sense any more. In fact, with autoPersist returning false, as e.g. the case for BlockedTicketSecretSyncAdapter, no data is saved any more!

Reply:
Agree. Now that the implementations of `insert`/`update` contain the DB queries, there is no need to distinguish on `BaseDownloadSyncAdapter` level. And with the current code it's very easy to make the mistake of defaulting to false like I did in the `BlockedTicketSecretSyncAdapter`

Reply:
Done